controllers/invoicestream.ctrl.go

```diff
@@ -0,0 +1,98 @@
+package controllers
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/getAlby/lndhub.go/common"
+	"github.com/getAlby/lndhub.go/db/models"
+	"github.com/getAlby/lndhub.go/lib/service"
+	"github.com/getAlby/lndhub.go/lib/tokens"
+	"github.com/gorilla/websocket"
+	"github.com/labstack/echo/v4"
+)
+
+// GetTXSController : GetTXSController struct
+type InvoiceStreamController struct {
+	svc *service.LndhubService
+}
+
+type InvoiceEventWrapper struct {
+	Type    string           `json:"type"`
+	Invoice *IncomingInvoice `json:"invoice,omitempty"`
+}
+
+func NewInvoiceStreamController(svc *service.LndhubService) *InvoiceStreamController {
+	return &InvoiceStreamController{svc: svc}
+}
+
+// Stream invoices streams incoming payments to the client
+func (controller *InvoiceStreamController) StreamInvoices(c echo.Context) error {
+	userId, err := tokens.ParseToken(controller.svc.Config.JWTSecret, (c.QueryParam("token")))
+	if err != nil {
+		return err
+	}
+	invoiceChan := make(chan models.Invoice)
+	subId := controller.svc.InvoicePubSub.Subscribe(userId, invoiceChan)
+	upgrader := websocket.Upgrader{}
+	upgrader.CheckOrigin = func(r *http.Request) bool { return true }
+	ticker := time.NewTicker(30 * time.Second)
+	ws, err := upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		controller.svc.InvoicePubSub.Unsubscribe(subId, userId)
+		return err
+	}
+	defer ws.Close()
+
+	//start listening for close messages
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		for {
+			_, _, err := ws.ReadMessage()
+			if err != nil {
+				return
+			}
+		}
+	}()
+
+	//start with keepalive message
+	err = ws.WriteJSON(&InvoiceEventWrapper{Type: "keepalive"})
```

Comment:
The ticker below only executes after 30 seconds. When writing a client, I found it easier if the websocket returned something immediately. So it's not strictly needed, but it can be useful to get a response immediately to really be sure that you are connected.